app/models/agents/s3_agent.rb

```diff
@@ -0,0 +1,206 @@
+    form_configurable :access_key_secret, roles: :validatable
+    form_configurable :region, type: :array, values: %w(us-east-1 us-west-1 us-west-2 eu-west-1 eu-central-1 ap-southeast-1 ap-southeast-2 ap-northeast-1 ap-northeast-2 sa-east-1)
+    form_configurable :watch, type: :array, values: %w(true false)
+    form_configurable :bucket, roles: :completable
+    form_configurable :filename
+    form_configurable :data_path
+
+    def validate_options
+      if options['mode'].blank? || !['read', 'write'].include?(options['mode'])
+        errors.add(:base, "The 'mode' option is required and must be set to 'read' or 'write'")
+      end
+      if options['bucket'].blank?
+        errors.add(:base, "The 'bucket' option is required.")
+      end
+      if options['region'].blank?
+        errors.add(:base, "The 'region' option is required.")
+      end
+
+      case interpolated['mode']
+      when 'read'
+        if options['watch'].blank? || ![true, false].include?(boolify(options['watch']))
+          errors.add(:base, "The 'watch' option is required and must be set to 'true' or 'false'")
+        end
+      when 'write'
+        if options['filename'].blank?
+          errors.add(:base, "filename must be specified in 'write' mode")
+        end
+        if options['data_path'].blank?
+          errors.add(:base, "data_path must be specified in 'write' mode")
+        end
+      end
+    end
+
+    def validate_access_key_id
+      !!buckets
+    end
+
+    def validate_access_key_secret
+      !!buckets
+    end
+
+    def complete_bucket
+      (buckets || []).collect { |room| {text: room.name, id: room.name} }
+    end
+
+    def working?
+      checked_without_error?
+    end
+
+    def check
+      return if interpolated['mode'] != 'read'
+      contents = safely do
+                   get_bucket_contents
+                 end
+      if boolify(interpolated['watch'])
+        watch(contents)
+      else
+        contents.each do |key, _|
+          create_event payload: get_file_pointer(key)
+        end
+      end
+    end
+
+    def get_io(file)
+      client.get_object(bucket: interpolated['bucket'], key: file).body
+    end
+
+    def receive(incoming_events)
+      return if interpolated['mode'] != 'write'
+      incoming_events.each do |event|
+        safely do
+          mo = interpolated(event)
+          client.put_object(bucket: mo['bucket'], key: mo['filename'], body: Utils.value_at(event.payload, mo['data_path']))
+        end
+      end
+    end
+
+    private
+
+    def safely
+      yield
+    rescue Aws::S3::Errors::AccessDenied => e
+      error("Could not access '#{interpolated['bucket']}' #{e.class} #{e.message}")
+    rescue Aws::S3::Errors::ServiceError =>e
+      error("#{e.class}: #{e.message}")
+    end
+
+    def watch(contents)
+      if last_check_at.nil?
+        self.memory = contents
+        return
+      end
+
+      new_memory = contents.dup
+
+      memory.each do |key, etag|
```

Comment:
You may want to nest this in the memory so we could add other top-level keys later.